### PR TITLE
TandemFoil Paper: 4L/192d depth at champion optimizer config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2949 (vash, sent back) found that 4L/192d is the optimal depth-width combination for TFP in a sweep using a different LR. Now test 4L/192d at the full champion config (Lion lr=1.25e-4, gc=0.5, EMA=0.999, T_max=10). The depth increase from 3L to 4L may provide better volume field modeling.

## Instructions

Change only `--model-layers 4` vs champion:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier \
  --model-layers 4 --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group tfp-arch-sweep
```

## Reporting Requirements
- Best val `field_mse` + epoch
- Test `field_mse` at best val checkpoint
- W&B run ID

## Baseline

**TFP baselines:**
- val best: `field_mse = 0.002383` @ ep443 (PR #3025, W&B: `d1xh0o1p`)
- Config: Lion lr=1.25e-4, T_max=10, gc=0.5, EMA=0.999, 3L/192d, Fourier+physics
- paper target: normalized full-field MSE (field_mse, lower is better)
- **No clean test_primary/field_mse yet** — that is what the TFP experiments must establish

Note: vash #2949 found 4L/192d best at different LR. haku #3056 is sweeping LR at 3L. This tests 4L at champion LR.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT robin:

## Results

### TFP 4L/192d at Champion Config — NEGATIVE RESULT

| Metric | Value | Baseline (3L/192d) |
|--------|-------|-------------------|
| val_primary/field_mse | **Infinity** (all 118 epochs) | 0.002383 @ ep443 |
| val/volume_mse_p | Infinity (never finite) | — |
| val/surface_mse_p (best) | 1.15e+19 @ ep71 | — |
| val/surface_mse_Ux (best) | 0.00156 @ ep114 | — |
| val/surface_mse_Uy (best) | 0.00551 @ ep111 | — |
| Train loss (best) | 0.051 @ ep111 | — |
| Divergence epoch | 116 | 462 |

**W&B run:** x982sfdv (robin/tfp-4l-192d-champion)
**Peak GPU memory:** ~54.7 GB (GPU 0)
**Epochs completed:** 118 (diverged at ep116, run still in progress but unrecoverable)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier \
  --model-layers 4 --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent robin --wandb-name "robin/tfp-4l-192d-champion" --wandb-group tfp-arch-sweep
```

## What happened

**The 4L/192d model at champion config completely failed to produce a finite field_mse in 118 epochs.** The experiment is a clear negative result — adding depth from 3L to 4L at champion config is harmful for TFP.

### Key observations:

1. **Volume pressure never went finite.** The `volume_mse_p` channel was Infinity for all 118 epochs. This is the fundamental blocker — `field_mse` includes volume metrics and thus stayed Infinity throughout.

2. **Surface pressure oscillated wildly.** It went from Infinity (ep1-17) → first finite value at ep18 (6e+27) → oscillated between 1e+19 and Infinity for the rest of training. Best surface pressure was 1.15e+19 at ep71, still ~21 orders of magnitude from baseline. The pressure predictions overflow the `sinh()` inverse transform.

3. **Velocity channels converged well.** Surface Ux reached 0.00156 and Uy reached 0.00551 — the model learns velocity representations fine. The issue is specifically the pressure channel.

4. **Training loss decreased normally.** Loss went from 1.05 to 0.051 in asinh-transformed space, proving the model IS learning in that space. The issue is that pressure predictions in physical space (after sinh inverse) overflow.

5. **Diverged at epoch 116** — T_max=10 cosine cycling instability hit at ep116 (grad norms: 12→24→75→174). The 3L baseline diverged at ep462 — the 4L model is 3.9x more susceptible to this instability.

### Why 4L fails where 3L succeeds:

The extra layer amplifies the magnitude of pressure predictions. The model uses `asinh(p * scale)` for training targets and `sinh(pred) / scale` for evaluation. With 4 layers, the model's pressure predictions in asinh space are more extreme, causing `sinh()` to overflow to Infinity during evaluation. The velocity channels don't undergo this transform and converge normally at both 3L and 4L.

The 4L model also hits the T_max=10 divergence instability ~4x earlier than 3L, suggesting the deeper model has a more fragile optimization landscape.

## Suggested follow-ups

1. **4L with lower LR (e.g., 8e-5 or 6e-5)** — the higher grad norms with 4L suggest the champion LR is too aggressive for the deeper model. Reducing LR might prevent the pressure overflow and the early divergence.
2. **4L with T_max=20 or T_max=30** — the less aggressive cosine cycling would reduce the instability that caused divergence at ep116.
3. **4L with gc=0.3** — tighter gradient clipping might prevent the pressure prediction magnitudes from growing too large.
4. **4L without asinh pressure transform** — the overflow is specifically in the sinh inverse. Disabling the transform would eliminate the overflow issue (but may hurt pressure learning quality).
5. **3L depth is optimal for TFP at champion config** — this result combined with vash's sweep (which found 4L best at different LR) suggests 4L needs config re-tuning, not a direct transfer from the 3L champion recipe.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3094 because ADVISOR: Closing — 4L/192d at champion config is a dead end on TFP. val_primary/field_mse was Infinity for all 140 epochs — pressure channel overflows via asinh transform. Velocity channels converge (Ux=0.00156, competitive), confirming the issue is pressure-specific. Diverged at ep116 vs ep462 for 3L champion — T_max=10 cycling is too aggressive for 4L. Good diagnosis, robin — 4L needs its own LR/T_max tuning, which is a separate hypothesis..

**Comment by tcapelle** [comment]
STUDENT robin:

## Update — Run completed

The run completed 432 epochs total (360-min timeout). After the divergence at ep116, the model never recovered — loss climbed monotonically from 0.051 (ep111 best) to 0.849 (ep432). field_mse remained Infinity throughout all 432 epochs. The divergence was permanent and catastrophic.

Final state: loss=0.849, field_mse=Inf, all pressure channels=Inf.
